src/libostree/ostree-repo-pull.c

```diff
@@ -1322,6 +1324,165 @@ commitstate_is_partial (OtPullData   *pull_data,
 }
 
 static gboolean
+ensure_collection_ref_in_summary (GVariant   *summary,
+                                  const char *collection_id,
+                                  const char *ref)
+{
+  g_assert_nonnull (collection_id);
+  g_assert_nonnull (ref);
+
+  g_autoptr(GVariant) additional_metadata = NULL;
+  additional_metadata = g_variant_get_child_value (summary, 1);
+
+  const char *main_collection_id;
+  if (!g_variant_lookup (additional_metadata,
+                         OSTREE_SUMMARY_COLLECTION_ID,
+                         "&s",
+                         &main_collection_id))
+    main_collection_id = NULL;
+
+  gboolean check_main_collection;
+  check_main_collection = (g_strcmp0 (main_collection_id, collection_id) == 0);
+
+  if (check_main_collection)
+    {
+      g_autoptr(GVariant) refs = NULL;
+      gsize i, n;
+
+      refs = g_variant_get_child_value (summary, 0);
+      for (i = 0, n = g_variant_n_children (refs); i < n; i++)
+        {
+          const char *refname;
+          g_autoptr(GVariant) ref_v = g_variant_get_child_value (refs, i);
+
+          g_variant_get_child (ref_v, 0, "&s", &refname);
+          if (g_strcmp0 (refname, ref) == 0)
+            return TRUE;
+        }
+
+      return FALSE;
+    }
+
+  g_autoptr(GVariant) collection_map = NULL;
+  collection_map = g_variant_lookup_value (additional_metadata,
+                                           OSTREE_SUMMARY_COLLECTION_MAP,
+                                           G_VARIANT_TYPE ("a{sa(s(taya{sv}))}"));
+  if (collection_map != NULL)
+    {
+      GVariantIter collection_map_iter;
+      char *unpack_summary_collection_id;
+      GVariant *unpack_refs;
+
+      g_variant_iter_init (&collection_map_iter, collection_map);
+
+      while (g_variant_iter_next (&collection_map_iter,
```

Comment:
Use `g_variant_iter_loop()` to get fancy memory handling of `unpack_summary_collection_id` and `unpack_refs`.